packages/@uppy/aws-s3-multipart/src/index.js

```diff
@@ -140,9 +140,14 @@ class HTTPCommunicationQueue {
   }
 
   async getUploadId (file, signal) {
-    const cachedResult = this.#cache.get(file.data)
-    if (cachedResult != null) {
-      return cachedResult
+    let cachedResult
+    while ((cachedResult = this.#cache.get(file.data)) != null) {
```

Comment:
how is it possible to have multiple values in the face for the same key?

Reply:
Not sure what you mean, there shouldn't be multiple values, the loop is there to dismiss the cached rejections (if the cache value is a promise that rejects, we request for a new key).

Reply:
Can you update this comment to tell the full story? Would be nice to add context where (`X`) we set the upload id and why it's safe to block?

```suggestion
    // As the cache is updated asynchronously, there could be a race condition where we just miss
    // a new result so we loop here until we get something back. It is safe to block until we get a result
    // because in `X` we do Y 
    while ((cachedResult = this.#cache.get(file.data)) != null) {
```

Reply:
```suggestion
    // As the cache is updated asynchronously, there could be a race condition
    // where we just miss a new result so we loop here until we get nothing back,
    // at which point it's out turn to create a new cache entry.
    while ((cachedResult = this.#cache.get(file.data)) != null) {
```

Reply:
We do, line 160

Reply:
oh wow I really missed that line 😅 👍